### PR TITLE
Add notifications concern methods to load notifications

### PR DESCRIPTION
**Pull request by xzhflying**

### CODE REVIEW

File: app/controllers/concerns/application_notifications_concern.rb [hunk 1]
```diff
@@ -11,4 +11,25 @@ def notification_view_path(notification)
     notification_class_name = notification.class.name.underscore.gsub('/', '_').pluralize
     "/#{root_path}/#{notification_class_name}/#{notification.notification_type}"
   end
+
+  # Loads recent activity feeds of a given course. Accessible in views using +@activity_feeds+
+  #
+  # @param [Course] course The target course.
+  # @param [Integer] count The count of activity feeds that will be loaded. By default it will
+  #   load all activity feeds.
+  def load_recent_activity_feeds(course, count = :unset)
```
Comment (allenwq) [inline app/controllers/concerns/application_notifications_concern.rb]: Move this directly under course/controller. It's not meant to be here.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -16,4 +16,7 @@ class ApplicationController < ActionController::Base
   include ApplicationThemingConcern
   include ApplicationUserConcern
   include ApplicationAnnouncementsConcern
+  include ApplicationNotificationsConcern
+
+  before_action { |c| c.load_unread_popups(current_user) }
```
Comment (allenwq) [inline app/controllers/application_controller.rb]: Why not put this in the concern ?

**Comment by allenwq** [comment]
@xzhflying  Where is the `render_notifications` ?

File: app/controllers/concerns/application_notifications_concern.rb [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
-  def notification_view_path(notification)
-    activity = notification.activity
-    root_path = "notifiers/#{activity.notifier_type}/#{activity.event}"
-    notification_class_name = notification.class.name.underscore.gsub('/', '_').pluralize
-    "/#{root_path}/#{notification_class_name}/#{notification.notification_type}"
+  included do
+    before_action :load_unread_popups
+  end
+
+  # Loads unread popups of current user. Accessible in views using +@popups+
```
Comment (lowjoel) [inline app/controllers/concerns/application_notifications_concern.rb]: don't use instance variables for globally accessible values.

File: app/controllers/concerns/application_notifications_concern.rb [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
+  def load_unread_popups
+    return [] if current_user.nil?
+    @popups = current_user.notifications.where(notification_type: 'popup').unread_by(current_user).
```
Comment (lowjoel) [inline app/controllers/concerns/application_notifications_concern.rb]: move this into a method on the model

File: app/controllers/course/controller.rb [hunk 5]
```diff
@@ -38,6 +38,18 @@ def current_component_host
                                                           self)
   end
 
+  # Loads recent activity feeds of a given course. Accessible in views using +@activity_feeds+
+  #
+  # @param [Course] course The target course.
+  # @param [Integer] count The count of activity feeds that will be loaded. By default it will
+  #   load all activity feeds.
+  def load_recent_activity_feeds(course, count = :unset)
```
Comment (lowjoel) [inline app/controllers/course/controller.rb]: this shouldn't be implemented for every single course controller.

Comment (lowjoel) [inline app/controllers/course/controller.rb]: actually this should be a concern and mixed into whoever needs it.

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -6,8 +6,7 @@ module ApplicationHelper
   include ApplicationAnnouncementsHelper
   include ApplicationWidgetsHelper
   include ApplicationCocoonHelper
-
-  include ApplicationNotificationsConcern
```
Comment (lowjoel) [inline app/helpers/application_helper.rb]: why the unnecessary whitespace change? :confused:

Comment (xzhflying) [inline app/helpers/application_helper.rb]: hmm... this whitespace is used to separate helpers and the concern before. But I made a mistake that all things should be helpers. So this whitespace is useless...

**Comment by xzhflying** [comment]
@lowjoel hmm...I've modified this pr two days ago. :flushed: Will this be OK?

File: app/controllers/concerns/application_notifications_concern.rb [hunk 7]
```diff
@@ -1,14 +1,16 @@
 module ApplicationNotificationsConcern
   extend ActiveSupport::Concern
-
-  # Returns the view path of the notification
+  # Loads recent activity feeds of a given course.
```
Comment (lowjoel) [inline app/controllers/concerns/application_notifications_concern.rb]: missing space before comment

File: app/controllers/concerns/application_notifications_concern.rb [hunk 8]
```diff
@@ -1,14 +1,16 @@
...
-  def notification_view_path(notification)
-    activity = notification.activity
-    root_path = "notifiers/#{activity.notifier_type}/#{activity.event}"
-    notification_class_name = notification.class.name.underscore.gsub('/', '_').pluralize
-    "/#{root_path}/#{notification_class_name}/#{notification.notification_type}"
+  # @param [Course] course The target course.
+  # @param [Integer] count The count of activity feeds that will be loaded. By default it will
+  #   load all activity feeds.
+  # @return [Array<CourseNotification>] Recent activity feed notifications
+  def load_recent_activity_feeds(course, count = :unset)
```
Comment (lowjoel) [inline app/controllers/concerns/application_notifications_concern.rb]: if this is for a course, this should not be included in ApplicationController.

File: app/helpers/application_notifications_helper.rb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+module ApplicationNotificationsHelper
+  # Returns the view path of the notification
+  #
+  # @param [Course::Notification|UserNotification] notification The target notification
```
Comment (lowjoel) [inline app/helpers/application_notifications_helper.rb]: is this a strict type requirement or a duck-type requirement?

File: app/models/user.rb [hunk 10]
```diff
@@ -71,6 +71,13 @@ def set_next_email_as_primary
     default_email_record.update_attributes(primary: true)
   end
 
+  # Loads unread popups of a user.
+  #
+  # @return [Array<UserNotification>] The unread popup notifications
+  def unread_popups
+    notifications.where(notification_type: 'popup').unread_by(self).order(created_at: :desc)
```
Comment (lowjoel) [inline app/models/user.rb]: Define a popups relation on the notification scope.

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -126,4 +126,22 @@
       end
     end
   end
+
+  describe '#unread_popups' do
+    let(:user) { create(:user) }
+    let!(:unread_popups) do
+      create(:user_notification, user: user, notification_type: :popup)
+    end
+    let!(:readed_popup) do
```
Comment (lowjoel) [inline spec/models/user_spec.rb]: `read_popup`

File: spec/models/user_spec.rb [hunk 12]
```diff
@@ -126,4 +126,22 @@
...
+      create(:user_notification, user: user, notification_type: :popup)
+    end
+
+    before do
+      readed_popup.mark_as_read! for: user
```
Comment (lowjoel) [inline spec/models/user_spec.rb]: just do this within the read_popup block.

File: app/helpers/application_notifications_helper.rb (hunk 9 above)
Comment (xzhflying) [inline app/helpers/application_notifications_helper.rb]: Actually it's a duck-type. As long as it is a kind of notification.

**Comment by xzhflying** [comment]
@lowjoel

File: spec/controllers/course/course_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+          controller.load_recent_activity_feeds(course).count
+        end
+
+        it 'returns all activity feeds' do
+          is_expected.to eq(2)
+        end
+      end
+    end
+  end
+end
```
Comment (xzhflying) [inline spec/controllers/course/course_controller_spec.rb]: This is solved.

File: app/controllers/concerns/application_notifications_concern.rb (hunk 8 above)
Comment (lowjoel) [inline app/controllers/concerns/application_notifications_concern.rb]: why is this still here?

File: app/helpers/application_notifications_helper.rb (hunk 9 above)
Comment (lowjoel) [inline app/helpers/application_notifications_helper.rb]: it doesn't matter if it's new or not, if it's a problem that we identify during a review, it should be fixed.

File: app/models/user.rb [hunk 14]
```diff
@@ -71,6 +71,13 @@ def set_next_email_as_primary
     default_email_record.update_attributes(primary: true)
   end
 
+  # Loads unread popups of a user.
+  #
+  # @return [Array<UserNotification>] The unread popup notifications
+  def unread_popups
+    notifications.popups.unread_by(self).order(created_at: :desc)
```
Comment (lowjoel) [inline app/models/user.rb]: I believe I said something about this line.

Comment (lowjoel) [inline app/models/user.rb]: this should probably be a concern. I don't think the user should know about the notification's internals.

File: app/controllers/concerns/application_notifications_concern.rb (hunk 8 above)
Comment (lowjoel) [inline app/controllers/concerns/application_notifications_concern.rb]: your spec is on the courses controller, why should this be in every single application controller?

Comment (xzhflying) [inline app/controllers/concerns/application_notifications_concern.rb]: uh? Maybe I got a wrong file name. This concern is just included in course controller.

File: app/models/user.rb (hunk 14 above)
Comment (xzhflying) [inline app/models/user.rb]: hmm... you mean re-put this in the concern?

File: app/controllers/concerns/course/activity_feeds_concern.rb [hunk 15]
```diff
@@ -0,0 +1,17 @@
+module Course::ActivityFeedsConcern
+  extend ActiveSupport::Concern
+
+  # Loads recent activity feeds of a given course.
+  #
+  # @param [Course] course The target course.
+  # @param [Integer] count The count of activity feeds that will be loaded. By default it will
+  #   load all activity feeds.
+  # @return [Array<CourseNotification>] Recent activity feed notifications
+  def load_recent_activity_feeds(course, count = :unset)
```
Comment (lowjoel) [inline app/controllers/concerns/course/activity_feeds_concern.rb]: if this is in the Course::Controller then you can just use `current_course`.

Comment (lowjoel) [inline app/controllers/concerns/course/activity_feeds_concern.rb]: and don't call it load_X, just call it X.

Comment (lowjoel) [inline app/controllers/concerns/course/activity_feeds_concern.rb]: and you shouldn't specify the count here, return a relation and just let the controller call limit()

File: app/models/user.rb (hunk 14 above)
Comment (lowjoel) [inline app/models/user.rb]: let me change this approach. @allenwq what do you think?

Comment (allenwq) [inline app/models/user.rb]: Move associations and this method to a model level concern ?

Comment (lowjoel) [inline app/models/user.rb]: whose concern is this? why does the user model know about the notifications' internals?

Comment (xzhflying) [inline app/models/user.rb]: Can this be a scope in user_notification model as `unread_popup`?

File: app/controllers/concerns/course/activity_feeds_concern.rb (hunk 15 above)
Comment (xzhflying) [inline app/controllers/concerns/course/activity_feeds_concern.rb]: Thanks! I'll do that right away.

File: app/models/user.rb (hunk 14 above)
Comment (lowjoel) [inline app/models/user.rb]: possible, but if you defined the popups scope, then you can merge it with the unread scope, there's no need for you to define a scope specifically.

**Comment by lowjoel** [comment]
Any updates? @xzhflying @allenwq

**Comment by xzhflying** [comment]
@lowjoel the rest problem is how to create an unread_notifications scope for a user.

File: app/controllers/concerns/course/activity_feeds_concern.rb (hunk 15 above)
Comment (lowjoel) [inline app/controllers/concerns/course/activity_feeds_concern.rb]: @xzhflying you didn't fix this?

File: app/models/user.rb (hunk 14 above)
Comment (lowjoel) [inline app/models/user.rb]: @xzhflying are you supposed to fix this?

Comment (lowjoel) [inline app/models/user.rb]: At the end of the day you removed this, why?

Comment (xzhflying) [inline app/models/user.rb]: This part is the unread scope problem that we have not solved. Also I've attached the link in slack. The link is about some discussions with the owner of 'unread' gem.

File: app/controllers/concerns/course/activity_feeds_concern.rb (hunk 15 above)
Comment (xzhflying) [inline app/controllers/concerns/course/activity_feeds_concern.rb]: hmm? This has been fixed.

File: app/models/user.rb (hunk 14 above)
Comment (lowjoel) [inline app/models/user.rb]: But it has nothing to do with this PR?

Comment (xzhflying) [inline app/models/user.rb]: hmm...I think I need to do something with that gem so that it can support our requirements. So maybe I'll fix this part in next pr.

File: app/controllers/concerns/course/activity_feeds_concern.rb [hunk 16]
```diff
@@ -0,0 +1,11 @@
+module Course::ActivityFeedsConcern
+  extend ActiveSupport::Concern
+
+  # Loads recent activity feeds of a given course.
+  #
+  # @return [Array<CourseNotification>] Recent activity feed notifications
+  def recent_activity_feeds
+    return [] if current_course.nil?
+    current_course.notifications.where(notification_type: 'feed').order(created_at: :desc)
```
Comment (lowjoel) [inline app/controllers/concerns/course/activity_feeds_concern.rb]: 'feed' should be a reference to a constant in the model.

**Comment by lowjoel** [comment]
@xzhflying Move the comments about the unread items to a separate PR, merge this first. Please rebase after fixing my latest comment.

**Comment by xzhflying** [comment]
Thanks, I've fixed that issue, and I think using the scope generated by 'enum' could be better. :flushed:  @allenwq

File: app/controllers/concerns/course/activity_feeds_concern.rb [hunk 17]
```diff
@@ -0,0 +1,11 @@
+module Course::ActivityFeedsConcern
+  extend ActiveSupport::Concern
+
+  # Loads recent activity feeds of a given course.
+  #
+  # @return [Array<CourseNotification>] Recent activity feed notifications
```
Comment (lowjoel) [inline app/controllers/concerns/course/activity_feeds_concern.rb]: Course::Notification right?

**Comment by lowjoel** [comment]
Yes, the scope is better.
